code/game/machinery/atmos_field_gen.dm

```diff
@@ -0,0 +1,289 @@
+		balloon_alert(user, "locked!")
+		return ITEM_INTERACT_FAILURE
+	return default_deconstruction_screwdriver(user, icon_state, icon_state, tool)
+
+/obj/machinery/atmos_shield_gen/crowbar_act(mob/user, obj/item/tool)
+	if(on)
+		balloon_alert(user, "turn off first!")
+		return ITEM_INTERACT_FAILURE
+	return default_deconstruction_crowbar(tool)
+
+/obj/machinery/atmos_shield_gen/wrench_act(mob/living/user, obj/item/tool)
+	if(on)
+		balloon_alert(user, "turn off first!")
+		return ITEM_INTERACT_FAILURE
+	if(locked)
+		balloon_alert(user, "unlock first!")
+		return ITEM_INTERACT_FAILURE
+	if(default_unfasten_wrench(user, tool) && !anchored)
+		turn_off()
+	return ITEM_INTERACT_SUCCESS
+
+/obj/machinery/atmos_shield_gen/attack_hand(mob/living/user, list/modifiers)
+	. = ..()
+	if(!anchored)
+		balloon_alert(user, "not anchored!")
+		return
+	if(locked)
+		balloon_alert(user, "locked!")
+		return
+	toggle(user)
+
+/obj/machinery/atmos_shield_gen/attack_hand_secondary(mob/user, list/modifiers)
+	. = ..()
+	if(. == SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN)
+		return
+	if(allowed(user))
+		locked = !locked
+		balloon_alert(user, "[locked ? "" : "un"]locked!")
+		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
+	balloon_alert(user, "no access!")
+	return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
+
+/obj/machinery/atmos_shield_gen/item_interaction(mob/living/user, obj/item/tool, list/modifiers)
+	. = NONE
+	add_fingerprint(user)
+	if(is_wire_tool(tool) && panel_open)
+		wires.interact(user)
+		return ITEM_INTERACT_SUCCESS
+	if(istype(tool, /obj/item/card/id) && check_access(tool))
+		locked = !locked
+		balloon_alert(user, "[locked ? "" : "un"]locked!")
+		return ITEM_INTERACT_SUCCESS
+
+/obj/machinery/atmos_shield_gen/process_early()
+	if(on == ACTIVE && !powered())
+		turn_off(power_failure = TRUE)
+	else if(on == ACTIVE_WANTPOWER && powered())
+		on = ACTIVE
+		update_appearance(UPDATE_OVERLAYS)
+
+	if(on != ACTIVE)
+		if(length(fields))
+			QDEL_LIST(fields)
+		return
+
+	if(!isnull(master))
+		if(master.on != ACTIVE) // FUCK
+			turn_off()
+		return
+	var/field_len = length(fields)
+	if(field_len)
+		use_energy(active_power_usage * field_len)
+		return
+	var/turf/current_turf = get_turf(src)
+	var/found_slave = null
+	for(var/i in 1 to max_range)
+		if(isnull(found_slave))
+			for(var/obj/machinery/atmos_shield_gen/generator in current_turf)
+				if(generator == src || !generator.powered() || generator.dir != REVERSE_DIR(dir) || !isnull(generator.master) || !generator.anchored)
+					continue
+				found_slave = generator
+				generator.on = ACTIVE
+				generator.update_appearance(UPDATE_OVERLAYS)
+				generator.master = src
+				break
+			if(!isnull(found_slave))
+				break
+		current_turf = get_step(current_turf, dir) // advance
+		if(isclosedturf(current_turf) || !isnull(locate(/obj/effect/atmos_shield) in current_turf)) // we were blocked by a wall or something
+			on = ACTIVE_WANTPOWER
+			update_appearance(UPDATE_OVERLAYS)
+			return
```

Comment:
can you explain to me what this code is doing?

Reply:
outdated idk what it did

Reply:
did you write it? was this like an old project?